src/treelearner/gpu_tree_learner.cpp

```diff
@@ -0,0 +1,1207 @@
+                                        *device_subhistograms_, *sync_counters_, device_histogram_outputs_);
+  }
+  Log::Info("%d dense features (%.2f MB) transfered to GPU in %f secs. %d sparse features.", 
+            dense_feature_map_.size(), ((dense_feature_map_.size() + (dword_features_ - 1)) / dword_features_) * num_data_ * sizeof(Feature4) / (1024.0 * 1024.0), 
+            end_time * 1e-3, sparse_feature_map_.size());
+  #if GPU_DEBUG >= 1
+  printf("Dense feature list (size %lu): ", dense_feature_map_.size());
+  for (i = 0; i < num_dense_features_; ++i) {
+    printf("%d ", dense_feature_map_[i]);
+  }
+  printf("\n");
+  printf("Sparse feature list (size %lu): ", sparse_feature_map_.size());
+  for (i = 0; i < num_features_ - num_dense_features_; ++i) {
+    printf("%d ", sparse_feature_map_[i]);
+  }
+  printf("\n");
+  #endif
+}
+
+void GPUTreeLearner::InitGPU(int platform_id, int device_id) {
+  // currently we don't use constant memory
+  int use_constants = 0;
+
+  // Get the max bin size, used for selecting best GPU kernel
+  max_num_bin_ = 0;
+  #if GPU_DEBUG >= 1
+  printf("bin size: ");
+  #endif
+  for (int i = 0; i < train_data_->num_features(); ++i) {
+    #if GPU_DEBUG >= 1
+    printf("%d, ", train_data_->FeatureNumTotalBin(i));
+    #endif
+    max_num_bin_ = std::max(max_num_bin_, train_data_->FeatureNumTotalBin(i));
+  }
+  #if GPU_DEBUG >= 1
+  printf("\n");
+  #endif
+  // initialize GPU
+  dev_ = boost::compute::system::default_device();
+  if (platform_id >= 0 && device_id >= 0) {
+    const std::vector<boost::compute::platform> platforms = boost::compute::system::platforms();
+    if ((int)platforms.size() > platform_id) {
+      const std::vector<boost::compute::device> platform_devices = platforms[platform_id].devices();
+      if ((int)platform_devices.size() > device_id) {
+        Log::Info("Using requested OpenCL platform %d device %d", platform_id, device_id);
+        dev_ = platform_devices[device_id];
+      }   
+    }   
+  }   
+  std::string kernel_source;
+  std::string kernel_name;
+  // determine which kernel to use based on the max number of bins
+  if (max_num_bin_ <= 16) {
+    kernel_source = kernel16_src_;
+    kernel_name = "histogram16";
+    device_bin_size_ = 16;
+    dword_features_ = 8;
+  }
+  else if (max_num_bin_ <= 64) {
+    kernel_source = kernel64_src_;
+    kernel_name = "histogram64";
+    device_bin_size_ = 64;
+    dword_features_ = 4;
+  }
+  else if ( max_num_bin_ <= 256) {
+    kernel_source = kernel256_src_;
+    kernel_name = "histogram256";
+    device_bin_size_ = 256;
+    dword_features_ = 4;
+  }
+  else {
+    Log::Fatal("bin size %d cannot run on GPU", max_num_bin_);
+  }
+  if(max_num_bin_ == 65) {
+    Log::Warning("Setting max_bin to 63 is sugguested for best performance");
+  }
+  if(max_num_bin_ == 17) {
+    Log::Warning("Setting max_bin to 15 is sugguested for best performance");
+  }
```

Comment:
When max_bin is set to 64, there can be 65 actual bins created (FeatureNumTotalBin() returns 65, as there is an extra "default bin"), so the GPU kernel with 256 bins has to be used. Similarly, when max_bin is set to 16, there can be 17 actual bins created, and GPU kernel with 64 kernels have to be used. When an user says "max_bin=64" (or "max_bin=16"), I think he/she wants to use the 64-bin (or 16-bin) GPU kernel for best performance. Thus I give a warning in case someone accidentally set max_bin to 64 or 16 (should be 63 or 15).